src/libopensc/card-rtecp.c

```diff
@@ -515,20 +526,25 @@ static int rtecp_change_reference_data(sc_card_t *card, unsigned int type,
 	/* put 0x80 TLV */
 	rsf_length[0] = (newlen >> 8) & 0xFF;
 	rsf_length[1] = newlen & 0xFF;
-	assert(buf_end - p >= (int)(2 + sizeof(rsf_length)));
+
+	if (buf_end - p < (int)(2 + sizeof(rsf_length)))
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
```

Comment:
`free(buf)` on error here and in the following `returns`. Sorry, missed this on the first round.

Reply:
Thanks for catching.